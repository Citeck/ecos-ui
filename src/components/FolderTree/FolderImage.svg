<svg height="15" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 29 22" fill="none">
  <path
  d="m0 3c0-1.7 1.3-3 3-3h6.5c1.2 0 2.3 0.7 2.7 1.8l0.2 0.4c0.5 1.1 1.6 1.8 2.7 1.8h10.9c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3h-23c-1.7 0-3-1.3-3-3v-16z"
  fill="#C6D6F0"
  />
</svg>
