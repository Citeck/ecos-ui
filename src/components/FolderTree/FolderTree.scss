@import '../../styles/constants';

.ecos-folder-tree {
  &__item {
    display: grid;
    grid-auto-flow: column;
    align-items: start;
    justify-content: start;
    grid-gap: 10px;
  }

  &__item-body {
    display: grid;
    grid-auto-flow: column;
    align-items: center;
    justify-content: start;
    grid-gap: 5px;
    cursor: pointer;
    height: 40px;
    user-select: none;

    &:hover {
      opacity: 0.7;
    }
  }

  &__item-title {
    font-size: $font-size-standard;
    font-weight: $font-weight-semi-bold;
    color: $text-color-dark;
    @extend %ellipsis;

    &_selected {
      font-weight: $font-weight-bold;
    }
  }

  &__folder-image {
    font-size: 15px;
    color: var(--light-primary-color);

    &_selected {
      color: var(--primary-color);
    }
  }

  &__fold-switch {
    height: 40px;
    display: grid;
    align-items: center;
    min-width: 15px;

    &-icon {
      color: $color-gray-5;
      font-size: 14px;
      cursor: pointer;
      width: 15px;
      min-width: 15px;
    }
  }
}
