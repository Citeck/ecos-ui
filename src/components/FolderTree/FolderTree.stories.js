import React from 'react';

import { demoItems, item1111 } from './__mocks__/FolderTree.mock';
import FolderTree from './FolderTree';

export default {
  title: 'Components/FolderTree',
  component: FolderTree,
  parameters: {
    actions: {
      argTypesRegex: '^on.*'
    }
  }
};

const defaultArgs = {
  items: demoItems
};

const Template = args => <FolderTree {...args} />;

export const Default = Template.bind({});
Default.args = {
  ...defaultArgs
};

export const Selected = Template.bind({});
Selected.args = {
  ...defaultArgs,
  selected: item1111.id
};
