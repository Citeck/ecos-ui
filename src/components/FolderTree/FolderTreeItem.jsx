import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import EcosIcon from '../common/EcosIcon';

import { ReactComponent as FolderImage } from './FolderImage.svg';

import './FolderTree.scss';

const FolderTreeItem = ({ children, item, isSelected, onSelect, onUnfold, onFold }) => {
  const { id, title, hasChildren, isUnfolded } = item;
  const _onSelect = () => typeof onSelect === 'function' && onSelect(id);
  const _onUnfold = () => typeof onUnfold === 'function' && onUnfold(id);
  const _onFold = () => typeof onFold === 'function' && onFold(id);

  return (
    <div className="ecos-folder-tree__item">
      <div className="ecos-folder-tree__fold-switch">
        {hasChildren && (
          <EcosIcon
            className="ecos-folder-tree__fold-switch-icon"
            data={{ value: isUnfolded ? 'icon-small-down' : 'icon-small-right' }}
            onClick={isUnfolded ? _onFold : _onUnfold}
          />
        )}
      </div>
      <div>
        <div className="ecos-folder-tree__item-body" onClick={_onSelect}>
          <FolderImage
            className={classNames('ecos-folder-tree__folder-image', {
              'ecos-folder-tree__folder-image_selected': isSelected
            })}
          />
          <span
            className={classNames('ecos-folder-tree__item-title', {
              'ecos-folder-tree__item-title_selected': isSelected
            })}
          >
            {title}
          </span>
        </div>
        <div className="ecos-folder-tree__item-children">{children}</div>
      </div>
    </div>
  );
};

const FolderTreeItemPropTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  parent: PropTypes.string,
  hasChildren: PropTypes.bool,
  isUnfolded: PropTypes.bool
};

FolderTreeItem.propTypes = {
  item: PropTypes.shape(FolderTreeItemPropTypes).isRequired,
  isSelected: PropTypes.bool,
  onSelect: PropTypes.func,
  onUnfold: PropTypes.func,
  onFold: PropTypes.func
};

export default FolderTreeItem;
export { FolderTreeItemPropTypes };
