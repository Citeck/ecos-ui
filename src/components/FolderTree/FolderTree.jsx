import React from 'react';
import PropTypes from 'prop-types';

import FolderTreeItem, { FolderTreeItemPropTypes } from './FolderTreeItem';
import './FolderTree.scss';

const FolderTree = ({ items, selected, onSelect, onUnfold, onFold }) => {
  const renderLevel = parent => {
    return items
      .filter(item => {
        if (!parent) {
          return !item.parent;
        }

        return item.parent === parent;
      })
      .map(item => {
        return (
          <FolderTreeItem
            key={item.id}
            item={item}
            isSelected={item.id === selected}
            onSelect={onSelect}
            onUnfold={onUnfold}
            onFold={onFold}
          >
            {renderLevel(item.id)}
          </FolderTreeItem>
        );
      });
  };

  return <div className="ecos-folder-tree">{renderLevel()}</div>;
};

FolderTree.defaultProps = {
  items: []
};

FolderTree.propTypes = {
  items: PropTypes.arrayOf(PropTypes.shape(FolderTreeItemPropTypes)).isRequired,
  selected: PropTypes.string,
  onSelect: PropTypes.func,
  onUnfold: PropTypes.func,
  onFold: PropTypes.func
};

export default FolderTree;
