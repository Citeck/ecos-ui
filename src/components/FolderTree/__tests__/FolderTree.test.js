import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import FolderTree from '../FolderTree';
import FolderTreeItem from '../FolderTreeItem';
import { item1, item1111, item1112, demoItems } from '../__fixtures__/FolderTree.fixtures';

describe('FolderTree tests', () => {
  describe('<FolderTree />', () => {
    it('should render FolderTree component', () => {
      const { container, asFragment } = render(<FolderTree />);
      expect(container).toMatchSnapshot();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render all visible items', () => {
      const { container } = render(<FolderTree items={demoItems} />);

      const visibleItems = demoItems.filter(item => {
        if (item.parent) {
          const parent = demoItems.find(i => i.id === item.parent);
          if (parent && !parent.isUnfolded) {
            return false;
          }
        }
        return true;
      });

      expect(container.getElementsByClassName('ecos-folder-tree__item')).toHaveLength(visibleItems.length);
    });
  });

  describe('<FolderTreeItem />', () => {
    it('should render FolderTreeItem component', () => {
      const { container, asFragment } = render(<FolderTreeItem item={{ id: '', title: '' }} />);
      expect(container).toMatchSnapshot();
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render title', () => {
      const { container } = render(<FolderTreeItem item={item1} />);
      const title = container.getElementsByClassName('ecos-folder-tree__item-title');
      expect(title).toHaveLength(1);
      expect(title[0].textContent).toBe(item1.title);
    });

    it('should render children', () => {
      const childText = 'Child text';
      const child = <div className="child-class">{childText}</div>;
      const { container } = render(<FolderTreeItem item={item1}>{child}</FolderTreeItem>);
      const children = container.getElementsByClassName('ecos-folder-tree__item-children');
      expect(children[0].getElementsByClassName('child-class')).toHaveLength(1);
      expect(children[0].textContent).toBe(childText);
    });

    it('should render selected', () => {
      const { container } = render(<FolderTreeItem item={item1} isSelected={true} />);
      expect(container.getElementsByClassName('ecos-folder-tree__folder-image_selected')).toHaveLength(1);
      expect(container.getElementsByClassName('ecos-folder-tree__item-title_selected')).toHaveLength(1);
    });

    it('should render switch icon if has children', () => {
      const { container } = render(<FolderTreeItem item={item1} />);
      const switchIcon = container.getElementsByClassName('ecos-folder-tree__fold-switch-icon');
      expect(switchIcon).toHaveLength(1);
    });

    it(`shouldn't render switch icon if has not children`, () => {
      const { container } = render(<FolderTreeItem item={item1111} />);
      const switchIcon = container.getElementsByClassName('ecos-folder-tree__fold-switch-icon');
      expect(switchIcon).toHaveLength(0);
    });

    it(`should render 'icon-small-down' switch icon if unfolded`, () => {
      const { container } = render(<FolderTreeItem item={item1} />);
      const switchIcon = container.getElementsByClassName('icon-small-down');
      expect(switchIcon).toHaveLength(1);
    });

    it(`should render 'icon-small-right' switch icon if folded`, () => {
      const { container } = render(<FolderTreeItem item={item1112} />);
      const switchIcon = container.getElementsByClassName('icon-small-right');
      expect(switchIcon).toHaveLength(1);
    });

    it(`should call 'onSelect' when click by item`, async () => {
      const user = userEvent.setup();

      const onSelect = jest.fn();
      const { container } = render(<FolderTreeItem item={item1} onSelect={onSelect} />);

      await user.click(container.getElementsByClassName('ecos-folder-tree__item-body')[0]);

      expect(onSelect).toHaveBeenCalledTimes(1);
      expect(onSelect).toHaveBeenCalledWith(item1.id);
    });

    it(`should call 'onUnfold' when click by folded switch icon`, async () => {
      const user = userEvent.setup();

      const onUnfold = jest.fn();
      const { container } = render(<FolderTreeItem item={item1112} onUnfold={onUnfold} />);

      await user.click(container.getElementsByClassName('ecos-folder-tree__fold-switch-icon')[0]);

      expect(onUnfold).toHaveBeenCalledTimes(1);
      expect(onUnfold).toHaveBeenCalledWith(item1112.id);
    });

    it(`should call 'onFold' when click by unfolded switch icon`, async () => {
      const user = userEvent.setup();

      const onFold = jest.fn();
      const { container } = render(<FolderTreeItem item={item1} onFold={onFold} />);

      await user.click(container.getElementsByClassName('ecos-folder-tree__fold-switch-icon')[0]);

      expect(onFold).toHaveBeenCalledTimes(1);
      expect(onFold).toHaveBeenCalledWith(item1.id);
    });
  });
});
