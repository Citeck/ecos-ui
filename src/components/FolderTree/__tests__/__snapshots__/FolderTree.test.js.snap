// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderTree tests <FolderImage /> should render FolderImage component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FolderImage />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <path
        d="m0 3c0-1.7 1.3-3 3-3h6.5c1.2 0 2.3 0.7 2.7 1.8l0.2 0.4c0.5 1.1 1.6 1.8 2.7 1.8h10.9c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3h-23c-1.7 0-3-1.3-3-3v-16z"
        fill="#C6D6F0"
      />,
      "className": undefined,
      "fill": "none",
      "viewBox": "0 0 29 22",
      "xmlns": "http://www.w3.org/2000/svg",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "d": "m0 3c0-1.7 1.3-3 3-3h6.5c1.2 0 2.3 0.7 2.7 1.8l0.2 0.4c0.5 1.1 1.6 1.8 2.7 1.8h10.9c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3h-23c-1.7 0-3-1.3-3-3v-16z",
        "fill": "#C6D6F0",
      },
      "ref": null,
      "rendered": null,
      "type": "path",
    },
    "type": "svg",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <path
          d="m0 3c0-1.7 1.3-3 3-3h6.5c1.2 0 2.3 0.7 2.7 1.8l0.2 0.4c0.5 1.1 1.6 1.8 2.7 1.8h10.9c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3h-23c-1.7 0-3-1.3-3-3v-16z"
          fill="#C6D6F0"
        />,
        "className": undefined,
        "fill": "none",
        "viewBox": "0 0 29 22",
        "xmlns": "http://www.w3.org/2000/svg",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "d": "m0 3c0-1.7 1.3-3 3-3h6.5c1.2 0 2.3 0.7 2.7 1.8l0.2 0.4c0.5 1.1 1.6 1.8 2.7 1.8h10.9c1.7 0 3 1.3 3 3v12c0 1.7-1.3 3-3 3h-23c-1.7 0-3-1.3-3-3v-16z",
          "fill": "#C6D6F0",
        },
        "ref": null,
        "rendered": null,
        "type": "path",
      },
      "type": "svg",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`FolderTree tests <FolderTree /> should render FolderTree component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FolderTree
    items={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [],
      "className": "ecos-folder-tree",
    },
    "ref": null,
    "rendered": Array [],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [],
        "className": "ecos-folder-tree",
      },
      "ref": null,
      "rendered": Array [],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`FolderTree tests <FolderTreeItem /> should render FolderTreeItem component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FolderTreeItem
    item={
      Object {
        "id": "",
        "title": "",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="ecos-folder-tree__fold-switch"
        />,
        <div>
          <div
            className="ecos-folder-tree__item-body"
            onClick={[Function]}
          >
            <FolderImage
              className="ecos-folder-tree__folder-image"
            />
            <span
              className="ecos-folder-tree__item-title"
            >
              
            </span>
          </div>
          <div
            className="ecos-folder-tree__item-children"
          />
        </div>,
      ],
      "className": "ecos-folder-tree__item",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "className": "ecos-folder-tree__fold-switch",
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="ecos-folder-tree__item-body"
              onClick={[Function]}
            >
              <FolderImage
                className="ecos-folder-tree__folder-image"
              />
              <span
                className="ecos-folder-tree__item-title"
              >
                
              </span>
            </div>,
            <div
              className="ecos-folder-tree__item-children"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <FolderImage
                  className="ecos-folder-tree__folder-image"
                />,
                <span
                  className="ecos-folder-tree__item-title"
                >
                  
                </span>,
              ],
              "className": "ecos-folder-tree__item-body",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "ecos-folder-tree__folder-image",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "",
                  "className": "ecos-folder-tree__item-title",
                },
                "ref": null,
                "rendered": "",
                "type": "span",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "ecos-folder-tree__item-children",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="ecos-folder-tree__fold-switch"
          />,
          <div>
            <div
              className="ecos-folder-tree__item-body"
              onClick={[Function]}
            >
              <FolderImage
                className="ecos-folder-tree__folder-image"
              />
              <span
                className="ecos-folder-tree__item-title"
              >
                
              </span>
            </div>
            <div
              className="ecos-folder-tree__item-children"
            />
          </div>,
        ],
        "className": "ecos-folder-tree__item",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
            "className": "ecos-folder-tree__fold-switch",
          },
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="ecos-folder-tree__item-body"
                onClick={[Function]}
              >
                <FolderImage
                  className="ecos-folder-tree__folder-image"
                />
                <span
                  className="ecos-folder-tree__item-title"
                >
                  
                </span>
              </div>,
              <div
                className="ecos-folder-tree__item-children"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <FolderImage
                    className="ecos-folder-tree__folder-image"
                  />,
                  <span
                    className="ecos-folder-tree__item-title"
                  >
                    
                  </span>,
                ],
                "className": "ecos-folder-tree__item-body",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "ecos-folder-tree__folder-image",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "",
                    "className": "ecos-folder-tree__item-title",
                  },
                  "ref": null,
                  "rendered": "",
                  "type": "span",
                },
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "ecos-folder-tree__item-children",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
